automatically map themes to groups

cezio opened the pull request:
implementation for fix #53

etj commented on ckanext/dcatapit/commands/dcatapit.py:
```diff
@@ -38,6 +39,22 @@ class DCATAPITCommands(CkanCommand):
+NAME_OF_THEM = list,of,groups,
+    separated by newline
+    or, by coma
+
```
Since we are no longer importing the mapping into the DB, this description should be removed.

etj commented on ckanext/dcatapit/model/dcatapit_model.py:
```diff
@@ -93,4 +94,134 @@ def persist(self, tag, lang):
+                        .first() or _get_group_from_session(gname)
+        if group is None:
+            group = Group(name=gname)
+            Session.add(group)
```
Please consider adding an option like `ckanext.dcatapit.theme_group_mapping.use_group` with allowed values:
- `only existing`: add the mapping only if the group already exists
- `force_create`: create the group if it doesn't exist

If the option isn't in the config file, the behaviour should be the most conservative one (`only existing` in this case)

etj commented on ckanext/dcatapit/model/dcatapit_model.py:
```diff
@@ -93,4 +94,134 @@ def persist(self, tag, lang):
+    """
+    fname = config.get(DCATAPIT_THEME_TO_MAPPING_SOURCE)
+    if not fname or not os.path.exists(fname):
+        log.warning("Cannot parse theme mapping, no such file: %s", fname)
```
Pls also consider if it's the case to create a brand new plugin inside the dcatapit extension (let's say `dcatapit_theme_group_mapper`?), which only implements the `after_create` and `after_update` methods. 
In this way, if the mapper functionality is wanted, the `dcatapit_theme_group_mapper` shall be added to the `ckan.plugins` line.

cezio commented:
> too many logs, allow this option to be turned off kindly

 I can just remove this log message. However, if deployment requires mapping, and admin will provide file, i think some warning should be emited if there's a possible misconfiguration

cezio commented:
>> too many logs, allow this option to be turned off kindly
> I can just remove this log message. However, if deployment requires mapping, and admin will provide file, i think some warning should be emited if there's a possible misconfiguration

Actually, maybe log msg should be emited only in case there's mapping in config, but path is non-existent.

cezio commented:
> Please address the 3 comments:
> oblsolete comment in the command line comands

 removed

> too many logs, allow this option to be turned off kindly

 Warning will be emited only if path doesn't exist

> allow groups not to be created automatically

added `ckanext.dcatapit.theme_group_mapping.add_new_groups` bool setting, defaults to `false`.

I've also added test for this functionality

etj commented:
Among the other, I have groups `econ1` and `econ2`

Mapping files contains:
```
[dcatapit:theme_group_mapping]

AGRI=agri-group
ECON=econ1,econ2
EDUC=
#ENER=
```

Created manually a dataset with theme ECON.
Had some problems with mandatory data, so the GUI presented the same UI several times before proceeding with the resource form.

I can read this message 3 times in the log (probably when I presses the save btn to go on to the resource input form).
```
2017-09-18 00:57:44,518 INFO  [ckanext.dcatapit.mapping] Read theme to groups mapping definition from /etc/ckan/theme_group_mapping.ini. 3 themes to map.
```

Got a dataset with the same groups repeated 3 times:
```
...
      "theme" : "ECON",   
...
      "groups" : [
         {
            "image_display_url" : "",
            "name" : "econ1",
            "id" : "92ee040c-2c1f-4e4e-b602-591501ac3e8f",
            "title" : "econ1",
            "description" : "",
            "display_name" : "econ1"
         },
         {
            "image_display_url" : "",
            "name" : "econ1",
            "id" : "92ee040c-2c1f-4e4e-b602-591501ac3e8f",
            "title" : "econ1",
            "description" : "",
            "display_name" : "econ1"
         },
         {
            "image_display_url" : "",
            "name" : "econ1",
            "id" : "92ee040c-2c1f-4e4e-b602-591501ac3e8f",
            "title" : "econ1",
            "description" : "",
            "display_name" : "econ1"
         },
         {
            "image_display_url" : "",
            "name" : "econ2",
            "id" : "aacff5a9-2b33-457b-8953-e0be427eca48",
            "title" : "econ2",
            "description" : "",
            "display_name" : "econ2"
         },
         {
            "image_display_url" : "",
            "name" : "econ2",
            "id" : "aacff5a9-2b33-457b-8953-e0be427eca48",
            "title" : "econ2",
            "description" : "",
            "display_name" : "econ2"
         },
         {
            "image_display_url" : "",
            "name" : "econ2",
            "id" : "aacff5a9-2b33-457b-8953-e0be427eca48",
            "title" : "econ2",
            "description" : "",
            "display_name" : "econ2"
         }
      ],
   ...
```
I guess we should check if a group is already associated to the current dataset.

cezio commented:
https://github.com/geosolutions-it/ckanext-dcatapit/pull/70#issuecomment-340625431

Yes, the code wasn't checking if group membership exists by default, also, case when multiple themes were mapped to overlapping groups wasn't handled well. Could you review this again?
